### ISSUE TITLE
Freqtrade UI: Multiple exit markers shown when using custom_exit (no populate_exit_trend)

### ISSUE THREAD

**Issue by PTRPlay** [freqtrade/freqtrade#12481]
## Describe your environment

  * Operating system: Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.36
  * Python Version: Python 3.13.8
  * CCXT version: 4.5.12
  * Freqtrade Version: freqtrade 2025.10
  

## Describe the problem:

When custom_exit() is the only exit logic and populate_exit_trend() is empty, Freqtrade UI plots multiple exit markers on the chart for a single trade. Please check the screenshot.
The strategy always exits 100% of the position, but the UI plots several exit points in a row.

Expected behavior:
Only one exit marker should be displayed for the trade.

When using populate_exit_trend, the exit markers are plotted normally.

My custom_exit() logic:
  ```
  def custom_exit(
        self,
        pair: str,
        trade: Trade,
        current_time: datetime,
        current_rate: float,
        current_profit: float,
        **kwargs,
    ) -> Optional[Union[str, bool]]:
       
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        if dataframe is None or dataframe.empty:
            return None

        last_candle = dataframe.iloc[-1]

        # Exit when price closes above previous high 2 times in a row
        if last_candle.get("consecutive_close_above_prev_high", 0) >= 2:
            return "consecutive_close_above_prev_high_2x"

        return None
```

(Only Candles and Trades are enabled on the chart)
<img width="2157" height="1105" alt="Image" src="https://github.com/user-attachments/assets/2b3324a0-90c2-4a43-bdc7-1f86475f97b4" />

**Comment by xmatthias** [comment]
Well based on a screenshot - there's little to nothing to go on to.
The UI plots the data it's provided - so we'll need to look at the actual data for details - which i'm pretty sure will reveal why it's plotting the way it is.


To look at this properly - please provide the following:
* does your strategy use `adjust_trade_position()`
* logs for the duration of one of the trades
* the output of `freqtrade show-trades --db-url sqlite:///<trades.sqlite> --print-json --trade-ids <tradeid>` for one of the trades with multiple rendered exits


Please use a "details block" for the show-trades output to keep the issue itself kindof readable (see below for details on how to do that). 

<details>
<summary>Expand to see how to do details blocks</summary>

```
<details>
<summary>Expand full </summary>
(wrap in tripple ticks `)
Trades output ...
</details>
```
</details>


**edit** just realized this seems to be backtesting? in that case - your best bet is to carefully look at the strategy (is the correct strategy used?? e.g. via `freqtrade list-strategies`).
in most of these cases - the bot was using a different strategy (version - from a different file with the same name) than what the user thought.

In any case - the UI will only show what the data contains - so the problem will be in your strategy - not with the visualization.
We'll need a proper reproduction for this (either your strategy and the commands you ran - or a reduced version of your strategy) - just one exit function won't do - as custom_exit alone can't cause partial exits.

**Comment by PTRPlay** [comment]
@xmatthias Thanks for the quick response. Here's more details:

Yes, the issue happened during backtesting.
I doubled check and I'm using correct strategy when backtesting.
I simplified the strategy as much as possible, and the issue remains reproducible. I am including the strategy and the command used to run it.

```
# ================================
# Backtesting command
# ================================
"""
docker compose run --rm freqtrade backtesting \
    --strategy FreqtradeUIIssue \
    --timerange 20250901- \
     -p BTC/USDT:USDT
"""


import numpy as np
import pandas as pd
import pandas_ta as ta
from datetime import datetime, timedelta, timezone
from pandas import DataFrame
from typing import Optional, Union

from freqtrade.strategy import (
    IStrategy,
    Trade,
    BooleanParameter,
    DecimalParameter,
    IntParameter
)

class FreqtradeUIIssue(IStrategy):
   
    INTERFACE_VERSION = 3
    can_short: bool = False

    minimal_roi = {
        "0": 0.03
    }

    stoploss = -0.08
    trailing_stop = False

    timeframe = "1h"

    process_only_new_candles = False

    use_exit_signal = True
    exit_profit_only = False
    ignore_roi_if_entry_signal = False

    startup_candle_count: int = 250


    # Plot configuration
    plot_config = {
        "main_plot": {
            "sma_20": {"color": "blue"},
            "sma_5": {"color": "orange"},
        }
    }


    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        dataframe['sma_5'] = dataframe['close'].rolling(window=5).mean()
        dataframe['sma_20'] = dataframe['close'].rolling(window=20).mean()

        dataframe["prev_candle_high"] = dataframe["high"].shift(1)

        # Track consecutive closes above previous high (needed for exit logic)
        dataframe["close_above_prev_high"] = (dataframe["close"] > dataframe["prev_candle_high"]).astype(int)
        dataframe["consecutive_close_above_prev_high"] = (
            dataframe["close_above_prev_high"]
            .groupby((dataframe["close_above_prev_high"] != dataframe["close_above_prev_high"].shift()).cumsum())
            .cumsum()
        )

        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
     
        conditions = (
           (dataframe["close"] > dataframe["sma_5"]) & (dataframe["close"] > dataframe["sma_20"])
        )
        
        dataframe.loc[conditions, "enter_long"] = 1
        dataframe.loc[conditions, "enter_tag"] = (
            f"above_both_smas"
        )

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        return dataframe

    def custom_exit(
        self,
        pair: str,
        trade: Trade,
        current_time: datetime,
        current_rate: float,
        current_profit: float,
        **kwargs,
    ) -> Optional[Union[str, bool]]:
       
        dataframe, _ = self.dp.get_analyzed_dataframe(pair, self.timeframe)
        if dataframe is None or dataframe.empty:
            return None

        last_candle = dataframe.iloc[-1]

        # Exit when price closes above previous high 2 times in a row
        if last_candle.get("consecutive_close_above_prev_high", 0) >= 2:
            return "consecutive_close_above_prev_high_2x"
            
        return None
```



**Update**: I'm really sorry for posting a stupid question. I found out that some trades were entered and exited immediately, and that caused the issue. On the chart it's hard to tell when this happens because in that case it only shows the percentage (for example, -0.1%) and there is no Long/Short label, so that's why I was confused. I’m closing the ticket. Thanks!

**Comment by xmatthias** [comment]
No worries - it's often difficult to see such things if you're deeply into your strategy (also without that, sometimes) 😆
